### ISSUE TITLE
error from ordinal model is improvable

### ISSUE THREAD

**Issue by tbates** [OpenMx/OpenMx#149]
This is the error from an ordinal ACE model that didn't converge adequately:

```R
model$output[['maxRelativeOrdinalError']] is larger than mvnRelEps ( 0.005 ) at the optimum.
 Standardized ordinal thresholds are too far from zero or you have too many ordinal variables with nonzero covariance.
 Increase the maximum number of integration points or reduce mvnRelEps.run model: perhaps mxTryHard would help?
```

would be good to point to a help page explaining all of this (what is too far from zero? what are nonzero covariances of thresholds? how many are typically too many? change max integration points how? to what? what are they now? Reduce mvnRelEps by how much? what is plausible or accepted for ordinal models? 

1. Tell people how to increase maximum number of integration points, and what is reasonable.
2. Tell people how to reduce mvnRelEps, and what is reasonable.
3. For ordinal models, suggest `mxTryHardOrdinal` (not mxTryHard)
3. Edit the warning "mvnRelEps.run model". Lack of a space in `Eps .run`. And should say  "Then re-run"? 
  (or is it just error message combining with other print out..?)

**Comment by jpritikin** [comment]
> what is too far from zero?

No idea.

> what are nonzero covariances of thresholds?

If a group of ordinal variables have no covariance with another group of ordinal variables then both groups can be integrated independently. Since this permis lower dimensional integration, accuracy and speed is generally improved. However, it is unlikely that this suggestion is actionable. If your ordinal variables are indicators in a factor model then they will have covariance and there is nothing you can do about it besides use an altogether different model.

> how many are typically too many?

It depends on current technology. You'll know that you have too many if integration takes forever or you get these complaints about accuracy.

> change max integration points how?

Yeah, we should point people to the mxOptions man page.

>  to what? what are they now?

The formula is in the mxOptions man page. Perhaps it help to explain that purpose of the integration points limit is prevent infinite looping during integration. I've never seen infinite looping; It's just a precaution.

> Reduce mvnRelEps by how much?

Divide by 5 or 10. It's fairly arbitrary.

> what is plausible or accepted for ordinal models?

It's hard to offer any guidance.

**Comment by mcneale** [comment]
I agree - it's difficult to say.   It's also difficult to say whether the number of evaluations should be increased (or to what number; doubling seems a reasonable approach).  Some amount of trial and error will be required.

**Comment by RMKirkpatrick** [comment]
As an aside...

> The formula is in the mxOptions man page. Perhaps it help to explain that purpose of the integration points limit is prevent infinite looping during integration. I've never seen infinite looping; It's just a precaution.

@jpritikin Is literally _infinite_ looping actually possible with SADMVN?  Or do you just mean _practically_ infinite looping?  My guess is the latter, but if it's the former, that sounds like a bug in SADMVN.

**Comment by jpritikin** [comment]
On Tue, Mar 05, 2019 at 07:22:44AM -0800, RMKirkpatrick wrote:
>    [1]@jpritikin Is literally infinite looping actually possible with
>    SADMVN? Or do you just mean practically infinite looping? My guess is
>    the latter, but if it's the former, that sounds like a bug in SADMVN.

I'm not sure. It didn't seem worth the effort trying to determine one or 
the other.

**Comment by mcneale** [comment]
I don't think it's possible for SADMVN to run indefinitely. I believe it has a maximum # of evaluations, and the conditions are whichever criterion is met first (abseps  or releps or maxevaluation).  But it sure can take a long time...

**Comment by tbates** [comment]
Here's a suggested warning:

```R
Just to note: The model finished with a larger relative ordinal error than we typically expect.
This may be fine depending on your model, but you may wish to re-run the model using
`mxTryHardOrdinal` in place of `mxRun` to try for a better fit.
Expert version: Technically model$output[['maxRelativeOrdinalError']] is 
larger than `mvnRelEps` ( 0.005 ) at the optimum. If this is expected for your model,
you might with to increase `mvnRelEps`, e.g:
mxOption(NULL, key="mvnRelEps", value= mxOption(NULL, key="mvnRelEps")*5)

see `?mxOptions`
```
Here's my rationale:

1. ordinal variables having nonzero covariance is what we expect in meaningful models and thresholds being too far from zero doesn't seem to have a defined meaning,  So we can delete both of these from the warning.
2. Maximum number of integration points is less useful, so we might not mention this?
4. Increasing mvnRelEPs will just make OpenMx treat the ~ .01 that it had reached  as `OK`? 
3. If anything don't we want to INCREASE (not decrease) mvnRelEps? (i.e., the problems running ordinal models arose from April 2018 decrease of mvnRelEps from .01 to .001, which was then backed off to .005.

Ergo: Perhaps we shouldn't ask users to muck with `mvnRelEPs` at all (or even mention it): Just say that the model may not have reached a global optimum and they should probably use `mxTryHardOrdinal` in place or mxRun to check...

Note all mxOptions says is:
>  mvnRelEps i  relative error tolerance

Not much use, I am guessing. Should this at least say "maximum tolerance for relative ordinal error"?

**Comment by jpritikin** [comment]
> If anything don't we want to INCREASE (not decrease) mvnRelEps? (i.e., the problems running ordinal models arose from April 2018 decrease of mvnRelEps from .01 to .001, which was then backed off to .005.

I would suggest increasing mvnRelEps unless the model is taking too long to evaluate. We generally want to make mvnRelEps closer to zero because that improves integration accuracy.

> Ergo: Perhaps we shouldn't ask users to muck with mvnRelEps at all (or even mention it): Just say that the model may not have reached a global optimum and they should probably use mxTryHardOrdinal in place or mxRun to check...

If an ordinal model involves very small thresholded areas then it might be necessary to reduce mvnRelEps toward zero to make the integration accurate enough.

**Comment by tbates** [comment]
Hmm... So that would mean allowing the model to run a lot longer as well, as it is already dieing having not reached the desired level of precision, no? So (assuming our default value of releps is right) we should tell people to allow the model to have more evaluations and point them to that?

**Comment by jpritikin** [comment]
>  So that would mean allowing the model to run a lot longer as well, as it is already dieing having not reached the desired level of precision, no?

Maybe. Perhaps it would help just to evaluate `model$output[['maxRelativeOrdinalError']]` instead of print how to extract that value out of the model output. Hm, but is `maxRelativeOrdinalError` actually diagnostic of anything? I'm not sure. If you pick a large `mvnRelEps` then `maxRelativeOrdinalError` might be fine but you could get completely wrong estimates because the integration error is too big.

**Comment by mcneale** [comment]
The SADMVN code says:
*            if INFORM = 1, completion with ERROR > EPS and MAXPTS
*                           function vaules used; increase MAXPTS to
*                           decrease ERROR;

I don't think MAXPTS can be set to +inf so it shouldn't run forever; maximum seems to be 2147483647 as it's a  4-byte integer.  The way we set options for integration is described in ?mxOption.  The formula is:
mvnMaxPointsA + mvnMaxPointsB*n + mvnMaxPointsC*n*n + exp(mvnMaxPointsD + mvnMaxPointsE * n * log(mvnRelEps))

so setting the following options:
mxOption(NULL, "mvnMaxPointsA", 2147483647) 
mxOption(NULL, "mvnMaxPointsB", 0) 
mxOption(NULL, "mvnMaxPointsC", 0) 
mxOption(NULL, "mvnMaxPointsD", 0) 
mxOption(NULL, "mvnMaxPointsE", 0) 

Should enable the maximum possible precision.  I note that having the defaults in the help page would be nice.  However, 2147483647 may still not be enough for an integral with, e.g., a very small value due to extreme threshold (e.g., where <.1% of endorse a binary item).  At that point, it may be worth considering, e.g., marginal maximum likelihood if the number of factors is small.

Of course, if the missingness is completely at random - and there are no essential moderators that would require FIML - then weighted least squares could be a good option.
